### ISSUE TITLE
[GraphQL] Subscriptions support

### ISSUE THREAD

**Issue by carlwoodhouse** [OrchardCMS/OrchardCore#4489]
Supporting subscriptions would allow us to notify consumers of the api that content has changed; which they might use for invalidating caching, rebuilding gatsby sites, etc.

more info:
* https://graphql-dotnet.github.io/docs/getting-started/subscriptions/
* https://github.com/graphql-dotnet/server/blob/develop/samples/Samples.Server/Startup.cs* 
https://graphql.org/blog/subscriptions-in-graphql-and-relay/

**Comment by MatthijsKrempel** [comment]
We have a simular request from the organisation, any pointers to how we should proceed?

**Comment by sebastienros** [comment]
You might be able to add a subscription already

Start with something like this `services.AddSingleton<ISchemaBuilder, ContentItemSubscription>();` which will give you access to the `ISchema` that contains access to queries, mutations and subscriptions. From that just follow the documentation from GraphQL.NET

**Comment by carlwoodhouse** [comment]
Yeah just follow what’s in the links; I think the only thing is you have to
enable web sockets .. plenty of examples about

**Comment by MatthijsKrempel** [comment]
Ok, not quite familiar with it, but I will have a look.
I don't think it is as simple as adding a schema, we also have to have the middleware handling the request.

**Comment by jrestall** [comment]
The sample link @carlwoodhouse provided is really good https://github.com/graphql-dotnet/server/blob/develop/samples/Samples.Server/Startup.cs

I would suggest implementing this under its own feature such as "OrchardCore.Apis.GraphQL.Subscriptions" so that WebSockets is only enabled when subscriptions are used. So you would call AddWebSockets/UseWebSockets from a SubscriptionsStartup class like how the Users module does it [here](https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Users/Startup.cs).

As for what the schema would look like for subscriptions, we need to discuss. Here's some initial options:

```
type Subscription {
  contentItem(where: SubscriptionWhereInput) : ContentItemSubscriptionPayload

  # OPTION 1
  blogPost(where: BlogPostSubscriptionWhereInput): BlogPostSubscriptionPayload

  # OPTION 2
  blogPost(where: SubscriptionWhereInput): BlogPostSubscriptionPayload

  # OPTION 3
  blogPostCreated : BlogPost
  blogPostUpdated: BlogPost
  blogPostRemoved: BlogPost
  blogPostPublished: BlogPost
...
}

# FILTER OPTION 1 (Simpler)
input SubscriptionWhereInput {
  # Filter for a specific mutation:
  # CREATED, UPDATED, REMOVED, PUBLISHED, UNPUBLISHED, CLONED
  mutation_in: [MutationType!]
}

# FILTER OPTION 2 (Allows for flexibility in future to filter on BlogPost specific fields.)
input BlogPostSubscriptionWhereInput {
  # Filter for a specific mutation:
  # CREATED, UPDATED, REMOVED, PUBLISHED, UNPUBLISHED, CLONED
  mutation_in: [MutationType!]
}

type BlogPostSubscriptionPayload {
  mutation: MutationType!
  node: BlogPost
}
```

Then create a ISchemaBuilder like `public class ContentTypeSubscriptions : ISchemaBuilder` similar to `ContentTypeQuery` that uses IObservable and ContentHandlerBase. Not sure exactly how this part would work currently.

We would actually be the first headless CMS to support graphql subscriptions once completed which would be cool! 🎉

**Comment by carlwoodhouse** [comment]
Totally agree it should be feature based due to the sockets dependency

**Comment by sebastienros** [comment]
I think you can only provide subs for 
- a specific content item, or a list of content items, by identifiers
- a list, or anything where we manage the collection (list part, taxonomy term)
- any content item, with a set of filters that can be computed in memory (predicates): content type, status, author, stereotype

Because a query (sql or lucene) would be too complex to handle.

I'd like to hear about scenarios for subscriptions, and see how these points would solve them.

**Comment by rserj** [comment]
>I'd like to hear about scenarios for subscriptions

Since Orchard Core an Application framework rather than just a CMS. Because of its multitenancy support, modularity, and great rendering engine, we have plans to use Orchard Core as FrontBack as a middleman between React client app and other backend services. We have non-CMS related scenarios where we need to send notifications from backend services to client React app, we use to use SignalR in the first Orchard for this purpose.

**Comment by rserj** [comment]
I dig a little bit further.
The problem is: Graphql-dotnet still relies on Single instance Schema. But we use ISchemaFactory.GetSchemaAsync<TSchema>()
we need it because if schema gets changed we can rebuild one plus each module can register their own ISchemaBuilder to modify the schema. 
We had such problem with single-instance schema before, please see
https://github.com/OrchardCMS/OrchardCore/issues/2164
https://github.com/graphql-dotnet/server/issues/107
It is still a problem with Subscriptions, their approach doesn't fit for ours, so we have to implement and support some services

These are places in Startup.cs where a single schema is used:
https://github.com/graphql-dotnet/server/blob/develop/samples/Samples.Server/Startup.cs

![image](https://user-images.githubusercontent.com/13991195/89476179-e2215180-d73e-11ea-896c-8d5cbdcec08a.png)
![image](https://user-images.githubusercontent.com/13991195/89476331-49d79c80-d73f-11ea-9e25-664314d6fb61.png)

I see multiple ways to adopt current implementation to our's
1) **We have to reimplement the next services to use our's factory approach for Subscriptions**

- GraphQLWebSocketsMiddleware<TSchema>
https://github.com/graphql-dotnet/server/blob/develop/src/Transports.Subscriptions.WebSockets/GraphQLWebSocketsMiddleware.cs

- WebSocketConnectionFactory<TSchema>
https://github.com/graphql-dotnet/server/blob/develop/src/Transports.Subscriptions.WebSockets/WebSocketConnectionFactory.cs


- DefaultGraphQLExecuter<TSchema>
https://github.com/graphql-dotnet/server/blob/develop/src/Core/Internal/DefaultGraphQLExecuter.cs
I have a concern about this one, cause it is under .Internal namespace. 


2) **Try to embrace it by using a hack with registration in our Startup.cs**

```
 services.AddTransient(typeof(IGraphQLExecuter<Schema>), provider =>
                    new DefaultGraphQLExecuter<Schema>(
                            provider.GetService<ISchemaFactory>().GetSchemaAsync().Result as Schema, 
                            provider.GetService<IDocumentExecuter>(),
                            provider.GetService<IOptions<GraphQLOptions>>(),
                            provider.GetServices<IDocumentExecutionListener>(),
                            provider.GetServices<IValidationRule>()));
```

3) **Another way to create a Decorator and register one as Single Instance schema like**

 ```
public class SchemaDecorator : Schema
    {
        public SchemaDecorator(IServiceProvider serviceProvider)
        {
            var dependencyResolver = serviceProvider.GetService<IDependencyResolver>();
            var factory = dependencyResolver.Resolve<ISchemaFactory>();
            var schema = factory.GetSchema().Result;
            .....// continue replace fields from our factory
```
and use it like in 
https://github.com/graphql-dotnet/server/blob/develop/samples/Samples.Server/Startup.cs
instead of ChatSchema use SchemaDecorator

4) **We can contact to graphql-dotnet dev and ask for solution/fix**

5) **Switch to Hot Chocolate https://hotchocolate.io here is a comparison**
https://www.reddit.com/r/graphql/comments/dpnqdx/hotchocolate_vs_graphqlnet/

cc: @sebastienros

**Comment by sebastienros** [comment]
Are you sure hotchocolate would solve the issue?

**Comment by carlwoodhouse** [comment]
I’d love to switch to hot chocolate; I’ve been using it away from ocore and
it’s a much better library; data loaders, Di support etc out of the box and
tons of other features. the perf is much better too.

However ..

Can’t say if it would fix this or not; as didn’t look into it.

The upgrade to ocore would be massive. I know this cos I started on a poc
branch several months ago.

It would be a major breaking change to anyone who has written their own
graph types etc. So would need serious thought.

**Comment by michaelstaib** [comment]
The upcoming version 11 solves your issues, as I understand. 

https://chillicream.github.io/hotchocolate/blog/2020/07/16/version-11

The new schema configuration API allows full async creation and update of dynamic schemas since Hot Chocolate supports schema stitching we reworked version 11 to be able to push schema changes to the GraphQL Gateway dynamically. Schema stitching/schema federation and standard schemas become the same and let you fuse multiple sources and update them on the fly. 

If you guys need help, let us now.

And sorry for barging in on this issue :) I just saw it linked to another issue on our project.

**Comment by sebastienros** [comment]
At that point I don't expect many have implemented custom graphql extensions. Breaking would be fine, as long as we document how to upgrade the code.

We'd need to check we can still provide the support we have done, with parts, where clauses, paging, ...

**Comment by rserj** [comment]
Unfortunately, I do not have time to implement Hot Chocolate for the whole Orchard.
But I implemented my own module based on the same abstractions we have, it was easy. I do not have to implement my own Middleware like we did with graphql-dotnet and rewriting SchemaFactory was pretty straightforward. 
I didn't integrate OrchardCore security model with hot chocolate but there are lots of integration points where we can do it.

I was thinking about using [Schema stitching](https://hotchocolate.io/docs/stitching), where each OrchardCore module/feature can define their own schema, and it will be up to the module/feature to decide either combine their schema with default single schema or expose module's schema as a separate end-point.
With Hot chocolate 11, it will be even easier to define schema right in the startup with a fluent builder api, please see
https://chillicream.github.io/hotchocolate/blog/2020/07/16/version-11

I can share my implementation with someone who wishes to start implementing Hot chocolate
